plugin/plugin.go

```diff
@@ -144,87 +146,110 @@ func (p *Plugin) OnTrafficFromClient(
+			if client != nil && client["remote"] != "" {
+				database, err = p.RedisClient.Get(ctx, client["remote"]).Result()
+				if err != nil {
+					CacheMissesCounter.Inc()
+					p.Logger.Debug("Failed to get cached response", "error", err)
+				}
+				CacheGetsCounter.Inc()
+			}
 		}
 
-		tables, err := postgres.GetTablesFromQuery(query)
-		if err != nil {
-			p.Logger.Debug("Failed to get tables from query", "error", err)
-			return resp, nil
+		// If the database is still not found, return the response as is without caching.
+		// This might also happen if the cache is cleared while the client is still connected.
+		// In this case, the client should reconnect and the error will go away.
+		if database == "" {
+			p.Logger.Debug("Database name not found or set in cache, startup message or plugin config. " +
+				"Skipping cache")
+			p.Logger.Debug("Consider setting the database name in the " +
+				"plugin config or disabling the plugin if you don't need it")
+			return
 		}
 
-		// Cache the table(s) used in each cached request. This is used to invalidate
-		// the cache when a rows is inserted, updated or deleted into that table.
-		for _, table := range tables {
-			requestQueryCacheKey := strings.Join([]string{table, cacheKey}, ":")
-			if err := p.RedisClient.Set(
-				ctx, requestQueryCacheKey, "", p.Expiry).Err(); err != nil {
+		cacheKey := strings.Join([]string{server["remote"], database, string(request)}, ":")
+		if errorResponse == "" && rowDescription != "" && dataRow != nil && len(dataRow) > 0 {
+			// The request was successful and the response contains data. Cache the response.
+			if err := p.RedisClient.Set(ctx, cacheKey, response, p.Expiry).Err(); err != nil {
 				CacheMissesCounter.Inc()
 				p.Logger.Debug("Failed to set cache", "error", err)
 			}
 			CacheSetsCounter.Inc()
+
+			// Cache the query as well.
+			query, err := postgres.GetQueryFromRequest(request)
+			if err != nil {
+				p.Logger.Debug("Failed to get query from request", "error", err)
+				return
+			}
+
+			tables, err := postgres.GetTablesFromQuery(query)
+			if err != nil {
+				p.Logger.Debug("Failed to get tables from query", "error", err)
+				return
+			}
+
+			// Cache the table(s) used in each cached request. This is used to invalidate
+			// the cache when a rows is inserted, updated or deleted into that table.
+			for _, table := range tables {
+				requestQueryCacheKey := strings.Join([]string{table, cacheKey}, ":")
+				if err := p.RedisClient.Set(
+					ctx, requestQueryCacheKey, "", p.Expiry).Err(); err != nil {
+					CacheMissesCounter.Inc()
+					p.Logger.Debug("Failed to set cache", "error", err)
+				}
+				CacheSetsCounter.Inc()
+			}
 		}
 	}
+}
 
+// OnTrafficFromServer is called when a response is received by GatewayD from the server.
+func (p *Plugin) OnTrafficFromServer(
+	_ context.Context, resp *v1.Struct,
+) (*v1.Struct, error) {
+	p.Logger.Info("Traffic is coming from the server side")
```

Comment:
Lower log level, as this might generate too many log lines and become noisy.
```suggestion
	p.Logger.Debug("Traffic is coming from the server side")
```